src/workerd/api/tracing-module.c++

```diff
@@ -0,0 +1,225 @@
+
+  if (IoContext::hasCurrent()) {
+    auto& ioContext = IoContext::current();
+    auto span = ioContext.makeUserTraceSpan(kj::mv(spanName));
+    auto impl = kj::heap<InternalSpanImpl>(kj::mv(span));
+    return js.alloc<InternalSpan>(ioContext.addObject(kj::mv(impl)));
+  } else {
+    // When no IoContext is available, create a no-op span
+    return js.alloc<InternalSpan>(kj::none);
+  }
+}
+
+jsg::JsValue TracingModule::startSpanWithCallback(jsg::Lock& js,
+    kj::String operationName,
+    jsg::Function<jsg::Value(jsg::Arguments<jsg::Value>)> callback,
+    jsg::Arguments<jsg::Value> args,
+    const jsg::TypeHandler<jsg::Ref<InternalSpan>>& jsSpanHandler,
+    const jsg::TypeHandler<jsg::Promise<jsg::Value>>& valuePromiseHandler) {
+
+  // Get the current span parent
+  if (!IoContext::hasCurrent()) {
+    // Create a resolver and promise
+    v8::Local<v8::Promise::Resolver> resolver;
+    if (!v8::Promise::Resolver::New(js.v8Context()).ToLocal(&resolver)) {
+      return js.undefined();
+    }
+    auto promise = resolver->GetPromise();
+
+    // Prepare callback arguments - create a new Arguments with span prepended
+    kj::Vector<jsg::Value> callbackArgVector;
+    // Add a no-op span as first argument
+    auto noOpSpan = js.alloc<InternalSpan>(kj::none);
+    callbackArgVector.add(js.v8Ref(jsSpanHandler.wrap(js, kj::mv(noOpSpan))));
+    for (auto& arg: args) {
+      callbackArgVector.add(arg.addRef(js));
+    }
+    auto callbackArgs = jsg::Arguments<jsg::Value>(callbackArgVector.releaseAsArray());
+
+    // Execute callback directly
+    auto result = js.tryCatch([&]() -> jsg::JsValue {
+      auto callbackResult = callback(js, kj::mv(callbackArgs));
+      return jsg::JsValue(callbackResult.getHandle(js));
+    }, [&](jsg::Value exception) -> jsg::JsValue {
+      resolver->Reject(js.v8Context(), exception.getHandle(js));
+      return jsg::JsValue(promise);
+    });
+
+    // If result is a promise, chain it to our resolver
+    v8::Local<v8::Value> resultHandle = result;
+    if (resultHandle->IsPromise()) {
+      auto resultPromise = resultHandle.As<v8::Promise>();
+      // For simplicity, just return the result promise directly
+      return result;
+    } else {
+      // Resolve with the result
+      resolver->Resolve(js.v8Context(), resultHandle);
+      return jsg::JsValue(promise);
+    }
+  }
+
+  SpanParent parent = IoContext::current().getCurrentTraceSpan();
+
+  // Create the span implementation
+  auto spanName = kj::ConstString(kj::str(operationName));
+
+  auto& ioContext = IoContext::current();
+  auto span = ioContext.makeUserTraceSpan(kj::mv(spanName));
+  auto impl = kj::heap<InternalSpanImpl>(kj::mv(span));
+  // Create the JavaScript span object
+  jsg::Ref<InternalSpan> jsSpan = js.alloc<InternalSpan>(ioContext.addObject(kj::mv(impl)));
+
+  // Create new span parent for the callback execution context
+  SpanParent newSpanParent = jsSpan->makeSpanParent();
```

Comment:
What is this variable supposed to do? A span parent does not affect the lifetime of the parent span.